src/Microsoft.AspNetCore.Routing/Matchers/ParameterSegmentMatchProcessor.cs

```diff
@@ -0,0 +1,65 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.AspNetCore.Http;
+
+namespace Microsoft.AspNetCore.Routing.Matchers
+{
+    internal sealed class ParameterSegmentMatchProcessor : MatchProcessor
+    {
+        private readonly int _segment;
+        private readonly string _name;
+        private readonly bool _isCatchAll;
+        private readonly bool _hasDefaultValue;
+        private readonly object _defaultValue;
+
+        public ParameterSegmentMatchProcessor(int segment, string name, bool isCatchAll)
+        {
+            _segment = segment;
+            _name = name;
+            _isCatchAll = isCatchAll;
+        }
+
+        public ParameterSegmentMatchProcessor(
+            int segment,
+            string name,
+            bool isCatchAll,
+            bool hasDefaultValue,
+            object defaultValue)
+        {
+            _segment = segment;
+            _name = name;
+            _isCatchAll = isCatchAll;
+            _hasDefaultValue = hasDefaultValue;
+            _defaultValue = defaultValue;
+        }
+
+        public override bool Process(
+            HttpContext httpContext,
+            string path,
+            ReadOnlySpan<PathSegment> segments,
+            RouteValueDictionary values)
+        {
+            if (segments.Length > _segment && _isCatchAll)
+            {
+                var segment = segments[_segment];
+                values[_name] = path.Substring(segment.Start, path.Length - segment.Start);
```

Comment:
Hmmm, has there ever been thought about making this Substring be lazy? The RVD keeps an array segment over the path and only does the Substring when the value is used by something.

I'm guessing not worth the complexity for such a small saving.

Reply:
This is an interesting idea, and I don't think one we've considered before. What would it look like though?

I'm trying to make this lazy as possible, but I haven't pushed that down to **inside** the RVD yet.